### PR TITLE
Allow attachments on meetings for agenda items

### PR DESCRIPTION
**Pull request by oliverguenther**
- [x] Introduces a meetings API so it can be plugged into the attachable mixin for uploading files, as agenda items are not journalized itself, but only in the meeting container
- [x] Extend the primer rich text input so that it can receive a resource

https://community.openproject.org/work_packages/26819

### CODE REVIEW

File: modules/meeting/app/forms/meeting_agenda_item/notes.rb [hunk 1]
```diff
@@ -27,15 +27,27 @@
...
+      }
     )
   end
 
   def initialize(disabled: false)
     @disabled = disabled
   end
+
+  def resource
+    return unless object&.meeting
```
Comment (dombesz) [inline modules/meeting/app/forms/meeting_agenda_item/notes.rb:48]: Just out of curiosity, in what case would we not have the meeting set?

File: modules/meeting/spec/lib/api/v3/meetings/meeting_representer_spec.rb [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+    end
+
+    it 'describes the object', :aggregate_failures do
+      expect(subject).to be_json_eql('Meeting'.to_json).at_path('_type')
+      expect(subject).to be_json_eql(meeting.id.to_json).at_path('id')
+      expect(subject).to be_json_eql(meeting.title.to_json).at_path('title')
+      expect(subject).to be_json_eql(meeting.lock_version.to_json).at_path('lockVersion')
+      expect(subject).to be_json_eql(meeting.start_time.utc.iso8601(3).to_json).at_path('startTime')
+      expect(subject).to be_json_eql(meeting.duration.to_json).at_path('duration')
+      expect(subject).to be_json_eql(meeting.location.to_json).at_path('location')
```
Comment (dombesz) [inline modules/meeting/spec/lib/api/v3/meetings/meeting_representer_spec.rb:94]: Just a minor note, `created_at` and `updated_at` fields are not tested.

File: modules/meeting/lib/api/v3/meetings/meeting_representer.rb [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+                 render_nil: true,
+                 getter: ->(*) {
+                   lock_version.to_i
+                 }
+
+        property :type,
+                 as: :meeting_type,
+                 getter: ->(*) { type }
+
+        date_time_property :start_time
```
Comment (dombesz) [inline modules/meeting/lib/api/v3/meetings/meeting_representer.rb:54]: Would it be useful to expose the `:end_time` attribute too? Probably as a readonly attribute.

File: docs/api/apiv3/components/schemas/meeting_model.yml [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+    description: Identifier of this meeting
+    readOnly: true
+    exclusiveMinimum: 0
+  title:
+    type: string
+    description: The meeting's title
+  _links:
+    type: object
+    properties:
+      addAttachment:
```
Comment (dombesz) [inline docs/api/apiv3/components/schemas/meeting_model.yml:48]: Shall we enumerate the other properties like `attachments`, `project`, `author`, `self`?

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+              value:
+                _embedded:
+                  elements:
+                  - _links:
+                      author:
+                        href: "/api/v3/users/1"
+                        title: OpenProject Admin
+                      container:
+                        href: "/api/v3/wiki_pages/72"
+                        title: wiki
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                        title: meeting
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 6]
```diff
@@ -0,0 +1,260 @@
...
+            response:
+              value:
+                _embedded:
+                  elements:
+                  - _links:
+                      author:
+                        href: "/api/v3/users/1"
+                        title: OpenProject Admin
+                      container:
+                        href: "/api/v3/wiki_pages/72"
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                        href: "/api/v3/meetings/72"
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 7]
```diff
@@ -0,0 +1,260 @@
+# /api/v3/wiki_pages/{id}/attachments
+---
+get:
+  parameters:
+  - description: ID of the wiki page whose attachments will be listed
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
  - description: ID of the meeting whose attachments will be listed
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 8]
```diff
@@ -0,0 +1,260 @@
+# /api/v3/wiki_pages/{id}/attachments
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
# /api/v3/meetings/{id}/attachments
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+                      raw: ''
+                    digest:
+                      algorithm: md5
+                      hash: 7ac9c97ef73d47127f590788b84c0c1c
+                    fileName: some.gif
+                    fileSize: 3521772
+                    id: 376
+                _links:
+                  self:
+                    href: "/api/v3/wiki_pages/72/attachments"
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                    href: "/api/v3/meetings/72/attachments"
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+          schema:
+            $ref: "../components/schemas/error_response.yml"
+          examples:
+            response:
+              value:
+                _type: Error
+                errorIdentifier: urn:openproject-org:api:v3:errors:NotFound
+                message: The requested resource could not be found.
+      description: |-
+        Returned if the wiki page does not exist or the client does not have sufficient permissions
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
        Returned if the meeting does not exist or the client does not have sufficient permissions
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+        to see it.
+
+        **Required permission:** view wiki pages
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
        **Required permission:** view meetings
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 12]
```diff
@@ -0,0 +1,260 @@
...
+
+        *Note: A client without sufficient permissions shall not be able to test for the existence of a work package.
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
        *Note: A client without sufficient permissions shall not be able to test for the existence of a meeting.
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 13]
```diff
@@ -0,0 +1,260 @@
...
+        That's why a 404 is returned here, even if a 403 might be more appropriate.*
+      headers: {}
+  tags:
+  - Attachments
+  description: ''
+  operationId: List_attachments_by_wiki_page
+  summary: List attachments by wiki page
+post:
+  parameters:
+  - description: ID of the wiki page to receive the attachment
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
  operationId: List_attachments_by_meeting
  summary: List attachments by meeting
post:
  parameters:
  - description: ID of the meeting to receive the attachment
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 14]
```diff
@@ -0,0 +1,260 @@
...
+                        href: "/api/v3/wiki_pages/72"
+                    _type: WikiPage
+                    id: 72
+                    title: wiki
+                _links:
+                  author:
+                    href: "/api/v3/users/1"
+                    title: OpenProject Admin
+                  container:
+                    href: "/api/v3/wiki_pages/72"
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                        href: "/api/v3/meeting/72/attachments"
                        method: post
                      attachments:
                        href: "/api/v3/meeting/72/attachments"
                      project:
                        href: "/api/v3/projects/12"
                        title: Demo project
                      self:
                        href: "/api/v3/meeting/72"
                    _type: Meeting
                    id: 72
                    title: meeting
                _links:
                  author:
                    href: "/api/v3/users/1"
                    title: OpenProject Admin
                  container:
                    href: "/api/v3/meeting/72"
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 15]
```diff
@@ -0,0 +1,260 @@
...
+              value:
+                _type: Error
+                errorIdentifier: urn:openproject-org:api:v3:errors:MissingPermission
+                message: You are not allowed to delete this attachment.
+      description: |-
+        Returned if the client does not have sufficient permissions.
+
+        **Required permission:** edit wiki pages
+
+        *Note that you will only receive this error, if you are at least allowed to see the wiki page*
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
        **Required permission:** edit meetings

        *Note that you will only receive this error, if you are at least allowed to see the meeting*
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 16]
```diff
@@ -0,0 +1,260 @@
...
+                _type: Error
+                errorIdentifier: urn:openproject-org:api:v3:errors:NotFound
+                message: The requested resource could not be found.
+      description: |-
+        Returned if the wiki page does not exist or the client does not have sufficient permissions
+        to see it.
+
+        **Required permission:** view wiki pages
+
+        *Note: A client without sufficient permissions shall not be able to test for the existence of a wiki page
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
        Returned if the meeting does not exist or the client does not have sufficient permissions
        to see it.

        **Required permission:** view meetings

        *Note: A client without sufficient permissions shall not be able to test for the existence of a meeting
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 17]
```diff
@@ -0,0 +1,260 @@
...
+
+        * Sending a file that is too large
+      headers: {}
+  tags:
+  - Attachments
+  description: |-
+    Adds an attachment with the wiki page as it's container.
+
+  operationId: Add_attachment_to_wiki_page
+  summary: Add attachment to wiki page
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
    Adds an attachment with the meeting as it's container.

  operationId: Add_attachment_to_meeting
  summary: Add attachment to meeting
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+                      addAttachment:
+                        href: "/api/v3/wiki_pages/72/attachments"
+                        method: post
+                      attachments:
+                        href: "/api/v3/wiki_pages/72/attachments"
+                      project:
+                        href: "/api/v3/projects/12"
+                        title: Demo project
+                      self:
+                        href: "/api/v3/wiki_pages/72"
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                        href: "/api/v3/meetings/72/attachments"
                        method: post
                      attachments:
                        href: "/api/v3/meetings/72/attachments"
                      project:
                        href: "/api/v3/projects/12"
                        title: Demo project
                      self:
                        href: "/api/v3/meetings/72"
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 19]
```diff
@@ -0,0 +1,260 @@
...
+                    _type: WikiPage
+                    id: 72
+                    title: wiki
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                    _type: Meeting
                    id: 72
                    title: meeting
```

File: docs/api/apiv3/paths/meeting_attachments.yml [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+                _links:
+                  author:
+                    href: "/api/v3/users/1"
+                    title: OpenProject Admin
+                  container:
+                    href: "/api/v3/wiki_pages/72"
+                    title: wiki
```
Comment (dombesz) [inline docs/api/apiv3/paths/meeting_attachments.yml]:
```suggestion
                    title: meeting
```

**Review by dombesz — CHANGES_REQUESTED** [review]
LGTM, just a few minor comments. I tried to check the documentation at the `/api/docs` and the newly added api information do not show up, probably we need to add them to the [dev/docs/api/apiv3/openapi-spec.yml](https://github.com/opf/openproject/blob/dev/docs/api/apiv3/openapi-spec.yml).

File: modules/meeting/app/forms/meeting_agenda_item/notes.rb (hunk 1 above)
Comment (oliverguenther) [reply modules/meeting/app/forms/meeting_agenda_item/notes.rb:48]: If you link a work package to a meeting, you get a new unpersisted agenda item without the meeting being set (as you can see the editor without the meeting being present).

This also means currently it's not possible to add attachments in that modal. I think that's fine for the first iteration though.

File: modules/meeting/lib/api/v3/meetings/meeting_representer.rb (hunk 3 above)
Comment (oliverguenther) [reply modules/meeting/lib/api/v3/meetings/meeting_representer.rb:54]: Yes good idea! 👍

File: docs/api/apiv3/components/schemas/meeting_model.yml (hunk 4 above)
Comment (oliverguenther) [reply docs/api/apiv3/components/schemas/meeting_model.yml:48]: Yes I missed this, thanks for spotting 👍

**Pull request merged**